dagster_polars/io_managers/delta.py

```diff
@@ -63,13 +67,26 @@ def dump_df_to_path(
 
         storage_options = self.get_storage_options(path)
 
-        df.write_delta(
-            str(path),
-            mode=context.metadata.get("mode") or self.mode,  # type: ignore
-            overwrite_schema=context.metadata.get("overwrite_schema") or self.overwrite_schema,
-            storage_options=storage_options,
-            delta_write_options=delta_write_options,
-        )
+        if context.metadata.get("mode") or self.mode != "merge":
+            df.write_delta(
+                str(path),
+                mode=context.metadata.get("mode") or self.mode,  # type: ignore
+                overwrite_schema=context.metadata.get("overwrite_schema") or self.overwrite_schema,
+                storage_options=storage_options,
+                delta_write_options=delta_write_options,
+            )
+        else:
+            ( 
+                df.write_delta(
+                    str(path),
+                    mode=context.metadata.get("mode") or self.mode,  # type: ignore
+                    storage_options=storage_options,
+                    delta_merge_options=delta_merge_options,
+                )
+                .when_matched_update_all()
```

Comment:
This needs to be configurable in some way, basically this is a default upsert, but MERGEs can be complex set of different update, delete and insert operations.

I commonly use deduplicate on insert

Reply:
Hi Ion,

You are right and if you look to my example you will find a deduplication strategy using a rank function over a primary key and then selecting the first row. However for that the input dataset needs to have a "cdc" column (like load dts).

Shouldn´t this be responsibility of the user?

An alternative could be to modify:

```python
 def get_metadata(self, context: OutputContext, obj: pl.DataFrame) -> Dict[str, MetadataValue]:
        assert context.metadata is not None

        metadata = super().get_metadata(context, obj)

        if context.has_asset_partitions:
            partition_by = context.metadata.get("partition_by")
            if partition_by is not None:
                metadata["partition_by"] = partition_by

        if context.metadata.get("mode") == "append":
            # modify the medatata to reflect the fact that we are appending to the table

            if context.has_asset_partitions:
                # paths = self._get_paths_for_partitions(context)
                # assert len(paths) == 1
                # path = list(paths.values())[0]

                # FIXME: what to about row_count metadata do if we are appending to a partitioned table?
                # we should not be using the full table length,
                # but it's unclear how to get the length of the partition we are appending to
                pass
            else:
                metadata["append_row_count"] = metadata["row_count"]

                path = self._get_path(context)
                # we need to get row_count from the full table
                metadata["row_count"] = MetadataValue.int(
                    DeltaTable(str(path), storage_options=self.get_storage_options(path))
                    .to_pyarrow_dataset()
                    .count_rows()
                )

        return metadata
```

To maybe do something like this:
```python

        if context.metadata.get("mode") == "append":
            # modify the medatata to reflect the fact that we are appending to the table

            if context.has_asset_partitions:
                # paths = self._get_paths_for_partitions(context)
                # assert len(paths) == 1
                # path = list(paths.values())[0]

                # FIXME: what to about row_count metadata do if we are appending to a partitioned table?
                # we should not be using the full table length,
                # but it's unclear how to get the length of the partition we are appending to
                pass
            else:
                metadata["append_row_count"] = metadata["row_count"]
       if context.metadata.get("mode") == "merge":
            # modify the medatata to reflect the fact that we are appending to the table
            metadata["primary_key"] == "something here that refers to this key"
            metadata["cdc_column"] == "something here that refers to this key"

                path = self._get_path(context)
                # we need to get row_count from the full table
                metadata["row_count"] = MetadataValue.int(
                    DeltaTable(str(path), storage_options=self.get_storage_options(path))
                    .to_pyarrow_dataset()
                    .count_rows()
                )

        return metadata


```

Reply:
Hey :)

Yeah, I need to go a bit more through the current implementation of dagster-polars. I've already pushed a PR for dagster-deltalake-polars as a first step.

Reply:
@edgBR for my own work I am planning to use the dagster-deltalake-polars and then only the parquet IO manager in dagster-polars. 

So somewhere next week after my first PR get's merged in dagster-deltalake-polars I will expand it there to cover a couple common MERGE strategies